update KEP 740 (#4852)
kubernetes/enhancements: keps/sig-auth/740-service-account-external-signing/README.md

Author: Done

Reviewer: Does  that get surfaced in the token header or claims, or in the jwks, or is it purely internal to kube-apiserver? if it gets surfaced to other token consumers / verifiers, then I don't think it should be part of this, at least to start

Reviewer: Discussed offline, this does surface in token headers, so will defer from this iteration, can capture in a "possible future work" section.

Definitely need to make sure the availability implications of an unavailable external signer on token issuance (and transitively pod startup) are well documented (here in the KEP/PRR section, but also in the website docs for folks considering whether to integrate with an external signer).

Author: Added  "Possible Future Work" section and Called it out there.

2. Enhanced security.
  kube-apiserver loads service account keys from a file on the disk. This means anyone with privileged access to the kube-apiserver configuration / filesystem could exfiltrate signing material. An external signer that never returns signing material would mitigate this risk.

### Goals

- Enable integration with HSMs and cloud KMSes
- Support out-of-process JWT signing
- Support listing public verifying keys
- Ensure tokens signed by external signers remain consistent with tokens issued by kube-apiserver (consistent headers, algorithms, claims, etc)
- Preserve existing behavior and performance for keys not read over a socket
<<[UNRESOLVED]>>
- Move legacy token signing from kube-controller-manager to kube-apiserver and thus external signing.
<<[/UNRESOLVED]>>

### Non-Goals

- Reading TLS serving certificates and key from a socket or reloading of kube-apiserver with new cert and key
- Reading any other certificates from a file

## Proposal

### User Stories

Allow kubernetes distributions to integrate with external JWT signers while preserving existing behavior. Choosing between existing and new behavior shall be configurable.

#### Preserve existing behavior

> The kube-apiserver flags `--service-account-key-file` and `--service-account-signing-key-file` will continue to be used for reading from files.

Author: Done

#### New API

> A new versioned grpc API (ExternalJWTSigner) will be created under `k8s.io/kubernetes/pkg/serviceaccount`. This will be similar to how the KMS envelope encryption has an API at `k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/v1beta1/service.proto`. It will have methods to sign JWTs and fetch supported keys (proto in [Design Details](#design-details) section). 

Reviewer 2: ```suggestion
A new versioned grpc API (ExternalJWTSigner) will be created under `k8s.io/kubernetes/pkg/serviceaccount`. This will be similar to how the KMS envelope encryption has an API at `k8s.io/kubernetes/staging/src/k8s.io/kms/apis/v2/api.proto`. It will have methods to sign JWTs and fetch supported keys (proto in [Design Details](#design-details) section). 
```

Author: Done

  // header cannot have any additional data that kube-apiserver does not recognize.
  // Already wrapped in URL-safe base64, exactly as it appears in the first segment of the JWT.
  string header = 1;

  // The signature for the JWT.
  // Already wrapped in URL-safe base64, exactly as it appears in the final segment of the JWT.
  string signature = 2;
}

message FetchKeysRequest {}

message FetchKeysResponse {
  repeated Key keys = 1;

  // The timestamp when this data was pulled from the authoritative source of
  // truth for verification keys.
  // kube-apiserver can export this from metrics, to enable end-to-end SLOs.
  google.protobuf.Timestamp data_timestamp = 2;

  // The timestamp for when it's appropriate to refetch verification keys to pick changes if any.
  // Will be treated as a hint to compute cache max age (`suggested_next_fetch_at.Sub(time.Now())/time.Second`).
  // max age will have a minimum floor of 10s (for both, kube-apiserver and OIDC provider) to avoid hot loop. 
  google.protobuf.Timestamp suggested_next_fetch_at = 3;
}

message Key {
  // A unique identifier for this key.
  string key_id = 1;

  // The public key, PKIX-serialized.
  bytes key = 2;

> }

Reviewer: Clusters that 1) move to external signers, 2) keep supporting legacy tokens (either via a static signer or a custom external signer controller), and 3) use a distinct signing key for legacy tokens _should_ be able to keep that legacy token signing key from showing up in oidc discovery (since legacy tokens aren't oidc-shaped and wouldn't be verifiable by oidc-discovery-doc consumers anyway)

To enable that, add an `exclude_from_oidc_discovery bool` to PublicKey to allow the external signer to return a key that can be used for verification but excluded from oidc discovery. Document it should only be set for verifying keys which are not used to issue bound tokens.

Author: Done

also called out the Legacy token scenario is a section ("Support for Legacy Tokens") under "Notes/Constraints/Caveats" section

message MetadataRequest {}

message MetadataResponse {
  // used by kube-apiserver for defaulting/validation of JWT lifetime while accounting for configuration flag values:
  // 1. `--service-account-max-token-expiration`
  // 2. `--service-account-extend-token-expiration`
  //
  // * If `--service-account-max-token-expiration` is greater than `max_token_expiration_seconds`, kube-apiserver treats that as misconfiguration and exits.
  // * If `--service-account-max-token-expiration` is not explicitly set, kube-apiserver defaults to `max_token_expiration_seconds`.
  // * If `--service-account-extend-token-expiration` is true, the extended expiration is `min(1 year, max_token_expiration_seconds)`.
  //
  // `max_token_expiration_seconds` must be at least 600s.
  int64 max_token_expiration_seconds = 1;
}

```

### Test Plan

[x] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

##### Unit tests

> Will need coverage for:

Author: Done

- pkg/kubeapiserver/options/authentication.go
- the signer plugin 
- the key cache plugin

<!-- TODO
Add info on concerned test in below form as read from https://testgrid.k8s.io/sig-testing-canaries#ci-kubernetes-coverage-unit

- `<package>`: `<date>` - `<test coverage>`
 -->

##### Integration tests

- Create a cluster with ExternalJWTSigner to configure and verify the health status of core components as published by kube-apiserver.

##### e2e tests

- Create a cluster with ExternalJWTSigner configured.
- Request a token for a service account principal.
- Use a token as bearer for making requests to kube-apiserver and ensure it succeeds.

### Graduation Criteria

#### Alpha

- Unit tests are completed
- Initial integration tests are completed with a dummy ExternalSigner.

#### Beta

- E2E tests are completed.
- We have at least one ExternalSigner implementation working with this change.

#### GA

- More than one ExternalSigner implementations are completed.
- Feature is tuned with feedback from distributions.

### Upgrade/Downgrade Strategy

Existing clusters preserving existing functionality: 
- Need no changes.

Existing clusters using enhanced functionality: 

> - Need a MasterVM workload that implements ExternalJWTSigner

Reviewer 2: ```suggestion
- Need a control plane plugin that implements ExternalJWTSigner
```

Author: Done

> - Need to configure the cluster with the `--service-account-signing-endpoint` flag. 

Reviewer 2: ```suggestion
- Need to configure the kube-apiserver with the `--service-account-signing-endpoint` flag. 
```

Author: Done

#### Different kube-apiserver versions running at the same time
- Skew is likely to exist when externalSigning is enabled/disabled in a live cluster.
- Distributions can take one of 2 approaches to handle it:
  1. Tolerate disruption
    - Disruption would exist until all kube-apiserver instances have not restarted with the new config (`--service-account-signing-endpoint` being set/unset).
    - During this, tokens can be issued by kube-apiserver running older version and fail auth when calling kube-apiserver running newer version.
    - The opposite is also true.
  2. Include old keys in supported set
    - This is for distributions that are looking for a smoother transition when enabling/disabling external singing on live clusters.
    - When disabling, distributions shall import keys from external signers to the on-disk key set.
    - When enabling, distributions shall implement an external signer such that it combines the external keys with on-disk keys. 

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

###### How can this feature be enabled / disabled in a live cluster?

> For distributions that require capability to enable the feature on a live cluster:

Reviewer 2: Need to mention setting the feature flag `ExternalServiceAccountTokenSigner` in the kube-apiserver (KAS)

Author: Done

> - ExternalJWTSigner needs to combine the supported keys that are listest from external key management solutions and the keys that were on the cluster before enablement.

Reviewer 2: ```suggestion
- ExternalJWTSigner needs to combine the supported keys that are listed from external key management solutions and the keys that were on the cluster before enablement.
```

Author: Done

> - Control plan will need a re-start with `--service-account-key-file` being un-set and `--service-account-signing-endpoint` being set.

Reviewer 2: ```suggestion
- kube-apiserver will need a restart with `--service-account-key-file` being un-set and `--service-account-signing-endpoint` being set.
```

what about the `--service-account-signing-key-file`? That needs to be unset as well?

Author: Done

> - former value of `--service-account-key-file` (a path on the file system) will need to be supplied to ExternalJWTSigner in some fashion(this is up to the owners of the distribution).

Author: Done

###### Does enabling the feature change any default behavior?

Not from an end-user perspective. 

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

It will be possible but will need additional work from respective distributions. Thus, each distribution can decide for themselves.

Can be achieved by satisfying the following:

> - Control plan will need a re-start with `--service-account-key-file` set and `--service-account-signing-endpoint` un-set.

Reviewer 2: ```suggestion
- kube-apiserver will need a restart with `--service-account-key-file` set and `--service-account-signing-endpoint` un-set.
```

Need to mention the `--service-account-signing-key-file` as well?

Author: Done

- Distributions will require importing the externally supported keys to the file system path as configured in `--service-account-key-file`.

> - If distributions do not intend to support tokens signed by external signers after the feature is disabled, then they can come up with a fair warning.

Reviewer 2: Add a note about setting `ExternalServiceAccountTokenSigner=false` in the KAS if they no longer want to keep the feature enabled.

Reviewer: I don't think they have to do anything with the feature gate, they would just leave `--service-account-signing-endpoint` unset

Author: Resolving per Jordan's comment

> - If distributions do not intend to support tokens signed by external signers after the feature is disabled, then they can come up with a fair warning.

Reviewer 2: Add a note about setting `ExternalServiceAccountTokenSigner=false` in the KAS if they no longer want to keep the feature enabled.

Author: same as https://github.com/kubernetes/enhancements/pull/4852#discussion_r1791106812

   - Check for `apiserver_request_total`
   - Use following labels: resource="serviceaccounts", subresource="token"
   - Compute error rate:

     ````
      sum(rate(apiserver_request_total{job="kubernetes-apiservers",group="", version="v1",resource="serviceaccounts",subresource="token",code=~"5.."}[5m]))
     ````

2. High usage of stale tokens being reported

   - Check for `serviceaccount_stale_tokens_total`
   - Compare against `serviceaccount_valid_tokens_total`

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

Will be covered by integration tests.

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

No.

### Monitoring Requirements

###### How can an operator determine if the feature is in use by workloads?

The Feature would not be used by workload directly but will be used by kube-apiserver.

The usage should be visible to the operator using Audit logs.
<!-- TODO 
  Add details on increasing audit log surface area for External signers
-->

###### How can someone using this feature know that it is working for their instance?

> The feature is not used by individual pods but the kube-apiserver itself. Simply successful Token requests and successful auth with kube-apiserver when using a bearer token in the request shall be indicators enough. 

Reviewer 3: Will we have a `/healthz/` endpoint for this feature?

Author: was discussed as part of https://github.com/kubernetes/enhancements/pull/4852#discussion_r1783239827

we decided to rely on status as reported by kube-apiserver since fetching JWKs falls in critical path.

Reviewer: the initial read of public keys should inform healthz / readyz, so there will be an endpoint for that initial sync ... once the first read happens successfully, subsequent failures wouldn't turn the API server unhealth, but they would be observable via metrics

Author: Also mentioned it in this section.

<!-- TODO
Needs Benchmarking on SLIs.
-->

###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?

- [x] Metrics
  - Metric name: `apiserver_request_total` and `apiserver_request_duration_seconds`
  - Aggregation method:  aggregate over `job="kubernetes-apiservers",group="", version="v1",resource="serviceaccounts",subresource="token"`
  - Components exposing the metric: kube-apiserver

###### Are there any missing metrics that would be useful to have to improve observability of this feature?

1. serviceaccount_jwks_freshness
2. serviceaccount_oidc_jwks_freshness
3. serviceaccount_jwks_fetch_requests_total

- 1 and 2 can both be gauge metrics. 
- Each will indicate how long it's been since the supported key set was last synced.
- 3 can be a cumulatively increasing metric with `status` that will also enable deriving error rate.

### Dependencies

One new dependency will be introduced and it will only be required for clusters configured/opted-in via the `--service-account-signing-endpoint` flag.

###### Does this feature depend on any specific services running in the cluster?

> The feature depends on a cluster level service that will implement [ExternalJWTSigner RPC](#externaljwtsigner-rpc) and will serve on UDS. each distribution will have their own version, so let's just address it as `ExternalJWTSigner`.

Reviewer 3: ```suggestion
The feature depends on a cluster level service that will implement [ExternalJWTSigner RPC](#externaljwtsigner-rpc) and will serve on a UDS. Each distribution will have their own version, so let's just address it as `ExternalJWTSigner`.
```

Author: Done

- ExternalJWTSigner
    - Usage description:
      This service will act as a client to the external key management solution that a distribution will integrate with.

      NOTE: this being a configurable/opt-in feature, the impact will only be seen on clusters that are using the feature.

      - Impact of its outage on the feature:
        - kube-apiserver will not be able to sign new service account JWTs.

>         - old JWTs will continue to be supported since we have an in cluster cache.

Reviewer 3: What will API server do if the cache expiration/suggested refresh time passes and can't be updated? Continue to use the old signing keys indefinitely?

Author: I want to say that It will continue operating with stale set of Keys. But now that I think about it, shall it have a TTL so we don't keep supporting a set of keys indefinitely? if so, how do we determine the TTL?

Reviewer: > What will API server do if the cache expiration/suggested refresh time passes and can't be updated? Continue to use the old signing keys indefinitely?

For verifying, yes.

new tokens signed by new `kid`s would fail if new public keys can't be fetched

Author: Corrected with mentions of 'existing keys will continue to be supported'.

- Sign:
  - This will be a call from kube-apiserver to an external service account JWT signer.
  - The call will be made every time kube-apiserver receives a service account token request.
  - It's a critical path to create pods and keep them running. 
  - QPS will vary according to the cluster usage.
  - ExternalJWTSigner's throughput will vary from distribution to distribution.

- FetchKeys:
  - This is again a call from kube-apiserver to the external service account JWT signer.
  - It will be a periodic call to keep the OIDC provider in sync with the supported keys; Frequency will be decided by `suggested_next_fetch_at` returned by an external signer. 
  - It will also be a dynamic call whenever kube-apiserver comes across a token signed by a key_id that it does not recognize.
  - QPS will be rather low since:
    - the signing keys shall not change very frequently.
    - even a small value of `suggested_next_fetch_at` will result in less than 1 qps.
  - ExternalJWTSigner's throughput will vary from distribution to distribution.

###### Will enabling / using this feature result in introducing new API types?

No new K8s api.

> ###### Will enabling / using this feature result in any new calls to the cloud provider?

Reviewer 3: By cloud provider do we mean external signing source? I'm thinking of on-prem HSMs that might otherwise be used

Author: I Agree, while one of the motivation is to enable off cluster keys management with different cloud providers, there is no preventing of what externalSigners can integrate with.

will changed the language using might/can.

  - all service account token requests will result in a call to a cloud-provider.
  - it might be less often depending on the implementation of `ExternalJWTSigner` in each individual distribution.

- Listing keys
  - There will be periodic calls from implementation of `ExternalJWTSigner` to cloud providers for syncing supported signing keys.
  - The volume of calls will depend on implementation of `ExternalJWTSigner` in each individual distribution.  

###### Will enabling / using this feature result in increasing size or count of the existing API objects?

No

###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?

Service account JWT signing will be delegated to external signers. So, `create token` calls which shall otherwise be compliant with [api_call_latency](https://github.com/kubernetes/community/blob/master/sig-scalability/slos/api_call_latency.md) SLO will now have a variable SLO that will be dictated by individual distribution's dependency(external signer).

###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?

No.

> In the context of dimensions specified in [supported limits], There will be 1 additional pod per master node. Additional resource usage will be subject to respective implementation. The increase should be negligible.

Reviewer 3: ```suggestion
In the context of dimensions specified in [supported limits], there will be 1 additional process (or pod) per control plane node. Additional resource usage will be subject to respective implementation. The increase should be negligible.
```

Author: Done

[supported limits]: https://git.k8s.io/community//sig-scalability/configs-and-limits/thresholds.md

###### Can enabling / using this feature result in resource exhaustion of some node resources (PIDs, sockets, inodes, etc.)?

not likely.

### Troubleshooting

<!-- TODO
This section must be completed when targeting beta to a release.

For GA, this section is required: approvers should be able to confirm the
previous answers based on experience in the field.

The Troubleshooting section currently serves the `Playbook` role. We may consider
splitting it into a dedicated `Playbook` document (potentially with some monitoring
details). For now, we leave it here.
-->

###### How does this feature react if the API server and/or etcd is unavailable?

feature is only accessible via kube-apiserver. JWT signing and authentication will anyways not work without kube-apiserver.

###### What are other known failure modes?

<!-- TODO
For each of them, fill in the following information by copying the below template:
  - [Failure mode brief description]
    - Detection: How can it be detected via metrics? Stated another way:

>       How can an operator troubleshoot without logging into a master or worker node?

Reviewer 3: ```suggestion
      How can an operator troubleshoot without logging into a control plane or worker node?
```

Author: Done

## Motivation

1. Ease of rotation.
  At present, kube-apiserver loads service account keys at process start time and the keys remain the same for the life-time of the process. Thus, rotating keys require a kube-apiserver to restart. External key management eliminates the need to restart.

2. Enhanced security.
  kube-apiserver loads service account keys from a file on the disk. This means anyone with privileged access to the kube-apiserver configuration / filesystem could exfiltrate signing material. An external signer that never returns signing material would mitigate this risk.

### Goals

- Enable integration with HSMs and cloud KMSes
- Support out-of-process JWT signing
- Support listing public verifying keys
- Ensure tokens signed by external signers remain consistent with tokens issued by kube-apiserver (consistent headers, algorithms, claims, etc)
- Preserve existing behavior and performance for keys not read over a socket

> - Move controller loop for legacy token signing to staging repo so it can be used for custom external signer implementation (depends on gathered f/b during Beta graduation)

Reviewer: ```
Move controller loop for legacy token signing to staging repo so it can be used for custom external signer implementation (depends on gathered f/b during Beta graduation)
```

I would drop this this from the overall goals of the KEP and leave it in the details / and the decision about it in the beta graduation requirement

Author: Done

- This static implementation can be replaced by a Dynamic one which integrates with ExternalJWTSigner if external signing is enabled.

#### Updates for supported public keys

- kube-apiserver holds reference to a static implementation of `ServiceAccountPublicKeysGetter` interface called `StaticPublicKeysGetter`.
- For clusters configured with an external signer, this static implementation will be replaced by a Dynamic one backed by a key cache (let's call it `ExternalPublicKeysGetter`).
- OIDC Discovery docs are served by `OpenIDMetadataServer` and it gets supported keys from the same `ServiceAccountPublicKeysGetter` as configured for kube-apiserver. 
- Thus, the OIDC endpoint will automatically serve external keys if the kube-apiserver is configured to use `ExternalPublicKeysGetter`. 

### Notes/Constraints/Caveats

- A new flag `--service-account-signing-endpoint` will be added to kube-apiserver specifying a unix socket where the key service will be accessible.
- The flag `--service-account-signing-endpoint` can either be set to the location of a UDS on a filesystem, or be prefixed with an @ symbol and name a UDS in the abstract socket namespace.

>    - Implementers shall note that using a filesystem based socket will weaken the pod isolation between kube-apiserver and external signer. That's because:

Author: Done

      - external has to run a root init container to set up the directory that will contain the socket on the host.
      - kube-apiserver has to have an external signer's group as a supplemental group in its security context.

>     - In contrast, the security model with abstract sockets is much cleaner. Any unix user can connect to the socket, but the plugin can get the UID and GID of the calling user and enforce its own access control based on that.

Author: Done

- combining local and external keys will not be supported inside the kube-apiserver. This *could* be achieved inside an external signer implementation by combining local public keys with remote public keys and remote signing requests to accommodate a migration.
- Specifying both, `--service-account-key-file/--service-account-signing-key-file` and `--service-account-signing-endpoint` shall result in an error.

#### Support for Legacy Tokens

> Implementers will have following options for legacy token support:

Author: Done

	1. Let the Controller loop run as it is with static signing keys. Stitch the public keys in external signer's JWKs.
	2. Turn off the loop (don't support legacy tokens) if external signing is enabled.

> 	3. Create a custom external signer for legacy tokens using Controller loop from staging repo (This option will only be available if demanded by Community as part of f/b for Beta graduation).

Author: Done

- **Risk:** With an external signer, token headers and claims can diverge from that in existing JWTs as signed by kube-apiserver.
  
  **Mitigation:** External signers will only return signature and header while kube-apiserver will assemble the JWT. This will allow kube-apiserver to validate the header and control the claims.

## Design Details

### ExternalJWTSigner RPC

Will be served on a socket as configured via the `--service-account-signing-endpoint` flag.

```proto
syntax = "proto3";

// This service is served by a process on a local Unix Domain Socket.
service ExternalJWTSigner {
  // Sign takes a serialized JWT payload, and returns the serialized header and
  // signature.  The caller can then assemble the JWT from the header, payload,
  // and signature.
  //
  // The plugin MUST set a key id in the returned JWT header.
  rpc Sign(SignJWTRequest) returns (SignJWTResponse) {}

  // FetchKeys returns the set of public keys that are trusted to sign
  // Kubernetes service account tokens. Kube-apiserver will call this RPC:
  //

>   // * Every time it tries to validate a JWT carrying an unknown key ID, and

Author: Done

  //
  // * Periodically, so it can serve reasonably-up-to-date keys from the OIDC
  //   JWKs endpoint.
  rpc FetchKeys(FetchKeysRequest) returns (FetchKeysResponse) {}
  
  // Metadata is meant to be called once on startup.
  // Enables sharing metadata with kube-apiserver (eg: the max token lifetime that signer supports)
  rpc Metadata(MetadataRequest) returns (MetadataResponse) {}
}

message SignJWTRequest {
  // URL-safe base64 wrapped payload to be signed.
  // Exactly as it appears in the second segment of the JWT

>   bytes claims = 1;

Author: Done

}

message SignJWTResponse {

>   // header must contain alg, kid, typ claims.

Author: Done

  // type must be “JWT”.
  // kid must be non-empty.
  // alg must be one of the algorithms supported by kube-apiserver (currently RS256, ES256, ES384, ES512).
  // header cannot have any additional data that kube-apiserver does not recognize.
  // Already wrapped in URL-safe base64, exactly as it appears in the first segment of the JWT.
  string header = 1;

  // The signature for the JWT.
  // Already wrapped in URL-safe base64, exactly as it appears in the final segment of the JWT.
  string signature = 2;
}

message FetchKeysRequest {}

message FetchKeysResponse {
  repeated Key keys = 1;

  // The timestamp when this data was pulled from the authoritative source of
  // truth for verification keys.
  // kube-apiserver can export this from metrics, to enable end-to-end SLOs.
  google.protobuf.Timestamp data_timestamp = 2;

  // refresh interval for verification keys to pick changes if any.
  // any value <= 0 is considered a misconfiguration.
  int64 refresh_hint_seconds = 3;
}

message Key {
  // A unique identifier for this key.
  string key_id = 1;

>   // The public key, PKIX-serialized.

Reviewer: same restrictions on key types as the algorithm restrictions in SignJWTResponse? 

must be a public key supported by kube-apiserver (currently RSA or ECDSA 256/384/521)?

Author: added a comment to document the same

  bytes key = 2;

>   // Set only for keys that are not used to sign bound tokens.

Author: added 

```
message SignJWTResponse {
 .
 .
 // kid must be non-empty and its corresponding public key should not be excluded from OIDC discovery.
 .
 .
 string header = 1;
}
```

and 

```
message Key {
 .
 .
 .
 // if set, external signer should not use this key to sign a JWT.
 bool exclude_from_oidc_discovery = 3;
}
```

  // eg: supported keys for legacy tokens.
  // If set, key is used for verification but excluded from OIDC discovery docs.
  bool exclude_from_oidc_discovery = 3;
}

message MetadataRequest {}

message MetadataResponse {
  // used by kube-apiserver for defaulting/validation of JWT lifetime while accounting for configuration flag values:
  // 1. `--service-account-max-token-expiration`
  // 2. `--service-account-extend-token-expiration`
  //
  // * If `--service-account-max-token-expiration` is greater than `max_token_expiration_seconds`, kube-apiserver treats that as misconfiguration and exits.
  // * If `--service-account-max-token-expiration` is not explicitly set, kube-apiserver defaults to `max_token_expiration_seconds`.

>   // * If `--service-account-extend-token-expiration` is true, the extended expiration is `min(1 year, max_token_expiration_seconds)`.

Reviewer 4: This seems like a weird interaction that will surprise people when it happens.

Author: Are you suggesting that it should rather result in an validation error?

cc: @liggitt

Reviewer: IIRC, @deads2k wanted the external signer to be able to control the ultimate expirations without the kube-apiserver args needing to change. The alternatives are:
1. fail to start the kube-apiserver
2. silently default `--service-account-extend-token-expiration` to false
3. try to get the external signer to sign year-long extended duration tokens anyway
4. make the extended token duration as long as possible while fitting in the bounds of the external signer

I think 4 is the least surprising

Reviewer 4: My comment was mostly around 'how do we make it clear to folks that tokens that used to be extended to a year now are extended to only a week' _before_ they have workloads that depend on the extension start breaking seemingly at random (i.e. the pod has been running long enough).

I agree that 4 seems like an okay option given the constraints.

Reviewer: I'd put that in the category of things the admin / integrator should figure out before switching to an external token signer that will force token lifetimes to be shorter. Good to make a note of to document the specific metrics they can check to determine if extended token lifetimes are being relied on in the cluster (the stale tokens metric)

Author: called it out in a new section under "monitoring requirements"  ("Other useful metrics that can assist with this feature")

  //
  // `max_token_expiration_seconds` must be at least 600s.
  int64 max_token_expiration_seconds = 1;
}

```

### Test Plan

[x] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

##### Unit tests

<!-- 
  coverage at https://testgrid.k8s.io/sig-testing-canaries#ci-kubernetes-coverage-unit
 -->

- `pkg/kubeapiserver/options` : `10-08-2024` - `67.2`
- `pkg/controlplane/apiserver/options` : `10-08-2024` - `57.1`

> - `pkg/externaljwtsigner/` : `10-08-2024` - `0.0`

Author: Done

##### Integration tests

> - Create a cluster with ExternalJWTSigner to configure and verify the health status of core components as published by kube-apiserver.

Author: Done

##### e2e tests

- Create a cluster with ExternalJWTSigner configured.
- Request a token for a service account principal.

> - Use a token as bearer for making requests to kube-apiserver and ensure it succeeds.

Reviewer 4: I would still expect an e2e like the encrypt-all-kind one to exercise all the wiring, but I would expect ~95% of the testing to be done via unit/integration tests.

### Graduation Criteria

#### Alpha

- Unit tests are completed
- Initial integration tests are completed with a dummy ExternalSigner.

#### Beta

- E2E tests are completed.

> - We have at least one ExternalSigner implementation working with this change.

Author: Done

#### GA

- More than one ExternalSigner implementations are completed.
- Feature is tuned with feedback from distributions.

### Upgrade/Downgrade Strategy

Existing clusters preserving existing functionality: 
- Need no changes.

Existing clusters using enhanced functionality: 

> - Need a control plane plugin that implements ExternalJWTSigner

Author: Done

- Need to configure the kube-apiserver with the `--service-account-signing-endpoint` flag. 

### Version Skew Strategy

The change will completely live in the Control plane and thus, there is no question of a skew with component interaction. The only potential for a skew arises with different kube-apiserver versions running at the same time

#### Different kube-apiserver versions running at the same time
- Skew is likely to exist when externalSigning is enabled/disabled in a live cluster.
- Distributions can take one of 2 approaches to handle it:
  1. Tolerate disruption
    - Disruption would exist until all kube-apiserver instances have not restarted with the new config (`--service-account-signing-endpoint` being set/unset).
    - During this, tokens can be issued by kube-apiserver running older version and fail auth when calling kube-apiserver running newer version.

>     - The opposite is also true.

Author: Done

>   2. Include old keys in supported set

Author: only suggesting this now

>     - This is for distributions that are looking for a smoother transition when enabling/disabling external singing on live clusters.

Author: Done

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

It will be possible but will need additional work from respective distributions. Thus, each distribution can decide for themselves.

Can be achieved by satisfying the following:
- kube-apiserver will need a restart with `--service-account-key-file`/`--service-account-signing-key-file` set and `--service-account-signing-endpoint` un-set.
- Distributions will require importing the externally supported keys to the file system path as configured in `--service-account-key-file`.
- If distributions do not intend to support tokens signed by external signers after the feature is disabled, then they can come up with a fair warning.

###### What happens if we re-enable the feature if it was previously rolled back?

experience would be the same as when enabling.

###### Are there any tests for feature enablement/disablement?

- Unit and E2E tests will be added.
- The tests would **not** focus on continuing to support the same key set when enabling/disabling but would rather focus on viability of enabling/disabling.

### Rollout, Upgrade and Rollback Planning

> Since it is an opt-in feature; upgrading existing clusters to versions with this feature shall be a no-op in context of the feature.

Reviewer 4: Being opt-in is not relevant.  These questions assume the user wants to use the feature.

Author: Removed this statement and filled required sub-sections with the assumptions that users will opt in.

###### How can a rollout or rollback fail? Can it impact already running workloads?

> Since it is an opt-in feature, existing clusters shall remain unaffected. However, since we are touching paths for JWT signing and authentication, if a bug slips in, it can disrupt the respective operation.

Reviewer 5: The idea of this question is to imagine situation where you are enabling the feature and turning it on (with the kube-apiserver flag). What can potentially go wrong in that case?

Reviewer 5: For example the external jwt signer can't run, what will happen?

Reviewer 4: (this needs to be addressed)

Author: Done

###### What specific metrics should inform a rollback?

1. High error rates on service account token creation.

   - Check for `apiserver_request_total`
   - Use following labels: resource="serviceaccounts", subresource="token"
   - Compute error rate:

     ````
      sum(rate(apiserver_request_total{job="kubernetes-apiservers",group="", version="v1",resource="serviceaccounts",subresource="token",code=~"5.."}[5m]))
     ````

2. High usage of stale tokens being reported

>    - Check for `serviceaccount_stale_tokens_total`

Author: Replaced with:

```
2. Dramatic increase in reported 401 status code.

   - Check for `apiserver_request_total`
   - Use following labels: resource="serviceaccounts", subresource="token"
   - Compute error rate:

     ````
      sum(rate(apiserver_request_total{job="kubernetes-apiservers",group="", version="v1",code=~"401"}[5m]))
     ````
```

Captures the same intent.

   - Compare against `serviceaccount_valid_tokens_total`

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

Will be covered by integration tests.

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

No.

### Monitoring Requirements

###### How can an operator determine if the feature is in use by workloads?

The Feature would not be used by workload directly but will be used by kube-apiserver.

> The usage should be visible to the operator using Audit logs.

Reviewer 4: Please be more specific.

<!-- TODO 
  Add details on increasing audit log surface area for External signers
-->

###### How can someone using this feature know that it is working for their instance?

- [ ] Events
  - Event Reason: 
- [] API .status
  - Condition name: 

>   - Other field: 

Author: Done

3. serviceaccount_jwks_fetch_requests_total

- 1 and 2 can both be gauge metrics. 
- Each will indicate how long it's been since the supported key set was last synced.
- 3 can be a cumulatively increasing metric with `status` that will also enable deriving error rate.

### Dependencies

One new dependency will be introduced and it will only be required for clusters configured/opted-in via the `--service-account-signing-endpoint` flag.

###### Does this feature depend on any specific services running in the cluster?

The feature depends on a cluster level service that will implement [ExternalJWTSigner RPC](#externaljwtsigner-rpc) and will serve on a UDS. Each distribution will have their own version, so let's just address it as `ExternalJWTSigner`.

- ExternalJWTSigner
    - Usage description:
      This service will act as a client to the external key management solution that a distribution will integrate with.

      NOTE: this being a configurable/opt-in feature, the impact will only be seen on clusters that are using the feature.

      - Impact of its outage on the feature:
        - kube-apiserver will not be able to sign new service account JWTs.

>         - refreshing supported key sets will not be possible. Already held keys will continue to be supported.

Reviewer 4: ```suggestion
        - refreshing supported key sets will not be possible. Already held keys will continue to be supported unless the API server is restarted.
```

Author: Done

>         - kube-apiserver is supposed to fetch JWKs everytime it sees a new kid that it does not recognize; Thus, every call that requires Auth will lead to an attempted refresh(single-flight to prevent duplicate concurrent calls).   

Author: Done

> ## Table of Contents

Author: Done

## Motivation

1. Ease of rotation.
  At present, kube-apiserver loads service account keys at process start time and the keys remain the same for the life-time of the process. Thus, rotating keys require a kube-apiserver to restart. External key management eliminates the need to restart.

2. Enhanced security.
  kube-apiserver loads service account keys from a file on the disk. This means anyone with privileged access to the kube-apiserver configuration / filesystem could exfiltrate signing material. An external signer that never returns signing material would mitigate this risk.

### Goals

- Enable integration with HSMs and cloud KMSes
- Support out-of-process JWT signing
- Support listing public verifying keys
- Ensure tokens signed by external signers remain consistent with tokens issued by kube-apiserver (consistent headers, algorithms, claims, etc)
- Preserve existing behavior and performance for keys not read over a socket
<<[UNRESOLVED]>>

> - Move legacy token signing from kube-controller-manager to kube-apiserver and thus external signing.

Reviewer: Still chewing on what to do about this bit. The interaction between the controller that produces non-expiring tokens and an external signer that can rotate out signing keys or limit the max token lifetime is pretty weird...

Folks using external signers have a couple reasonable options I see:
1. Turn off the legacy token controller loop and choose not to support non-expiring legacy tokens in their cluster
2. a) Populate legacy token secrets with externally signed tokens *and* b) augment token controller logic with some awareness of when the tokens need to be replaced in those secrets.

1 is easy and can be accomplished by not giving kube-controller-manager a signing key.

2a _could_ be done by constructing the existing token controller loop with an alternate TokenGenerator implementation that talks to an external signer, 2b requires new bookkeeping and rotation logic.

For 2, I'm on the fence about whether it is worth it to try to teach kube-controller-manager about external signers and token rotation, or if most external signers will turn off legacy tokens or have their own rotation logic that would make a generic kube-controller-manager implementation less useful to them.

Reviewer: > I think we should move the existing controller to a staging repo, have it take an interface for the signer that can be implemented with any external signer, and leave it at that. That gives three choices to users:
>
> 1. Run the built-in signer with a file based signing key that is never rotated
> 2. Don't run the built-in signer or any custom re-implementation
> 3. Don't run the built-in signer and instead run a custom re-implementation using the staging repo

For option 1, would that mean their external signer impl would have to stitch in that public key for verification in its returned keyset?

2 seems ideal for clusters which can handle dropping legacy tokens capability

3 seems ~ok to me for clusters which cannot handle dropping legacy tokens capability

Reviewer: > I think we should move the existing controller to a staging repo, have it take an interface for the signer that can be implemented with any external signer, and leave it at that.

I'd be on board with that.

Reviewer: For alpha, settled on not doing anything in-tree for the legacy token controller, and documenting the options for folks using external signers as:
1. Turn off legacy token support
2. Keep legacy token controller in `kube-controller-manager` running using a file-based signing key and have the external signer return the legacy public key (optionally with `exclude_from_oidc_discovery: true`)
3. Implement and run a custom external controller to sign legacy tokens using an external signer

Add a beta graduation criteria to get feedback from folks who would choose option 3 on whether they would reuse legacy token controller code if it were located to a staging repo and allowed to take a token signer interface. If feedback indicates this would be valuable to the community, relocate that code as part of the beta graduation criteria.

Author: called out the Legacy token scenario is a section ("Support for Legacy Tokens") under "Notes/Constraints/Caveats" section

<<[/UNRESOLVED]>>

### Non-Goals

- Reading TLS serving certificates and key from a socket or reloading of kube-apiserver with new cert and key
- Reading any other certificates from a file

## Proposal

### User Stories

> Allow kubernetes distributions to integrate with external JWT signers while preserving existing behavior. Choosing between existing and new behavior shall be configurable via. flag(s).

Author: Done

#### Preserve existing behavior

The kube-apiserver flags `--service-account-key-file` and `--service-account-signing-key-file` will continue to be used for reading from files.

#### New API

A new versioned grpc API (ExternalJWTSigner) will be created under `k8s.io/kubernetes/pkg/serviceaccount`. This will be similar to how the KMS envelope encryption has an API at `k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/v1beta1/service.proto`. It will have methods to sign JWTs and fetch supported keys (proto in [Design Details](#design-details) section). 

#### Plugins

- A signer plugin will be a client to the new grpc API thus providing support for external signing.
- A public key cache will integrate with the same grpc to fetch supported public keys.

#### Updates for token generation

- The Kubernetes Control Plane config accepts `ServiceAccountIssuer` which is an interface of type `serviceaccount.TokenGenerator`
- `serviceaccount.TokenGenerator` is defined in `pkg/serviceaccount/jwt.go`.

> - Current implementation of `serviceaccount.TokenGenerator` is `serviceaccount.JWTTokenGenerator` which uses a static key for sining.

Author: Done

- This static implementation can be replaced by a Dynamic one which integrates with ExternalJWTSigner if external signing is enabled.

#### Updates for supported public keys

- kube-apiserver holds reference to a static implementation of `ServiceAccountPublicKeysGetter` interface called `StaticPublicKeysGetter`.
- For clusters configured with an external signer, this static implementation will be replaced by a Dynamic one backed by a key cache (let's call it `ExternalPublicKeysGetter`).
- OIDC Discovery docs are served by `OpenIDMetadataServer` and it gets supported keys from the same `ServiceAccountPublicKeysGetter` as configured for kube-apiserver. 
- Thus, the OIDC endpoint will automatically serve external keys if the kube-apiserver is configured to use `ExternalPublicKeysGetter`. 

### Notes/Constraints/Caveats

- A new flag `--service-account-signing-endpoint` will be added to kube-apiserver specifying a unix socket where the key service will be accessible.

> - The flag `--service-account-signing-endpoint` can either be set to the location of a UDS on a filesystem, or be prefixed with an @ symbol and name a UDS in the abstract socket namespace.

Reviewer 6: It is quite difficult to securely set up a filesystem-based socket between two static pods that are running as different unix users.  We have done it inside GKE for the KMS plugin integration, but the way we do it does meaningfully weaken the isolation between the kube-apiserver pod and the kmsplugin pod:
* kmsplugin has to run a root init container to set up the directory that will contain the socket on the host.
* kube-apiserver has to have kmsplugin's group as a supplemental group in its security context.

In contrast, the security model with abstract sockets is much cleaner.  Any unix user can connect to the socket, but the plugin can get the UID and GID of the calling user and enforce its own access control based on that.

However the strongest argument for supporting abstract sockets is that they "just work" when you use the golang networking APIs --- the behavior Harshal is describing for this flag is the native behavior of *net.UnixAddr.  We would have to write extra code to break abstract socket support.

> - The flag `--service-account-signing-endpoint` can either be set to the location of a UDS on a filesystem, or be prefixed with an @ symbol and name a UDS in the abstract socket namespace.

Reviewer 4: SGTM, lets include that reasoning in the KEP, especially since a plugin using an abstract socket needs to do its own ACL.

> - The flag `--service-account-signing-endpoint` can either be set to the location of a UDS on a filesystem, or be prefixed with an @ symbol and name a UDS in the abstract socket namespace.

Author: Done

- combining local and external keys will not be supported inside the kube-apiserver. This *could* be achieved inside an external signer implementation by combining local public keys with remote public keys and remote signing requests to accommodate a migration.

> - Specifying both, `--service-account-key-file` and `--service-account-signing-endpoint` shall result in an error.

Author: Done

### Risks and Mitigations

- New token generation and validation could suffer a performance difference due to communication over a socket and a potential remote RPC call (will vary for different signer implementations).

> - New token generation and validation could suffer a performance difference when reading over a socket, as an external process will be signing data.

Author: Done

> - Signing and verifying tokens over a grpc API carries the risk of a server side request forgery, where a malicious client could generate tokens. To mitigate this risk, the API will only be accessible over a unix socket protected by standard file permissions. Documentation will be added to ensure that distributions limit the socket to privileged or specific uids.

Author: Done

## Design Details

### ExternalJWTSigner RPC

Will be served on a socket as configured via the `--service-account-signing-endpoint` flag.

```proto
syntax = "proto3";

// This service is served by a process on a local Unix Domain Socket.
service ExternalJWTSigner {
  // Sign takes a serialized JWT payload, and returns the serialized header and
  // signature.  The caller can then assemble the JWT from the header, payload,
  // and signature.
  //
  // The plugin MUST set a key id in the returned JWT header.
  rpc Sign(SignJWTRequest) returns (SignJWTResponse) {}

  // FetchKeys returns the set of public keys that are trusted to sign
  // Kubernetes service account tokens. Kube-apiserver will call this RPC:
  //
  // * Every time it tries to validate a JWT carrying an unknown key ID, and
  //
  // * Periodically, so it can serve reasonably-up-to-date keys from the OIDC
  //   JWKs endpoint.
  rpc FetchKeys(FetchKeysRequest) returns (FetchKeysResponse) {}

  <<[UNRESOLVED]>>
  
  // Status is meant to be polled for external signer's health status.
  rpc Status(StatusRequest) returns (StatusResponse) {}
  
  <<[/UNRESOLVED]>>

> }

Author: Have added the Metadata() RPC with comments on intended usage.

message SignJWTRequest {
  // URL-safe base64 wrapped payload to be signed.
  // Exactly as it appears in the second segment of the JWT
  bytes claims = 1;
}

message SignJWTResponse {
  // header must contain alg, kid, typ claims.

>   // type must be “JWT”.

Author: Done

>   // type must be “JWT”.

Author: Done

  // alg must be one of the algorithms supported by kube-apiserver (currently RS256, ES256, ES384, ES512).

>   // Already wrapped in URL-safe base64, valid for directly inserting into the final JWT.

Author: Done

  bytes header = 1;

  // The signature for the JWT.  Already wrapped in URL-safe base64, valid for

>   // directly inserting into the final JWT.

Author: Done

>   bytes signature = 2;

Reviewer 4: Since these fields are base64, should this be a `string`?

Author: changed both, header and signature to string values.

}

message FetchKeysRequest {}

message FetchKeysResponse {
  repeated Key keys = 1;

  // The timestamp when this data was pulled from the authoritative source of
  // truth for verification keys.
  // kube-apiserver can export this from metrics, to enable end-to-end SLOs.
  google.protobuf.Timestamp data_timestamp = 2;

  // The timestamp for when it's appropriate to refetch verification keys to 
  // pick changes if any.
  google.protobuf.Timestamp suggested_next_fetch_at = 3;
}

message Key {
  // A unique identifier for this key.
  string key_id = 1;

  // The public key, PKIX-serialized.
  bytes key = 2;
}

> <<[UNRESOLVED]>>

Author: Removed Status RPC. Have not added anything to the metrics section yet but will capture it as I define monitoring requirements when targeting Beta to a release.

> <<[UNRESOLVED]>>

Author: Have added mentions to `Are there any missing metrics that would be useful to have to improve observability of this feature?` section.

message StatusRequest {}

message StatusResponse {
    // Any value other than "ok" is failing healthz.  
    string healthz = 2;
}

<<[/UNRESOLVED]>>
```

### Test Plan

[x] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

##### Unit tests

Will need coverage for:

- pkg/kubeapiserver/options/authentication.go
- the signer plugin 
- the key cache plugin

<!-- TODO
Add info on concerned test in below form as read from https://testgrid.k8s.io/sig-testing-canaries#ci-kubernetes-coverage-unit

- `<package>`: `<date>` - `<test coverage>`
 -->

##### Integration tests

- Create a cluster with ExternalJWTSigner to configure and verify the health status of core components as published by kube-apiserver.

##### e2e tests

- Create a cluster with ExternalJWTSigner configured.
- Request a token for a service account principal.
- Use a token as bearer for making requests to kube-apiserver and ensure it succeeds.

### Graduation Criteria

> <!-- TODO -->

Author: Added some bits.

For each stage, I am thinking (for externalSigners):
1. begin mock with
2. have atleast one real implementation.
3. have more than one implementations.

#### Alpha

#### Beta

#### GA

### Upgrade/Downgrade Strategy

Existing clusters preserving existing functionality: 
- Need no changes.

Existing clusters using enhanced functionality: 
- Need a MasterVM workload that implements ExternalJWTSigner
- Need to configure the cluster with the `--service-account-signing-endpoint` flag. 

### Version Skew Strategy

> N/A (The change will completely live in Control plane)

Author: Have called it out along with available options to handle it.

###### Does enabling the feature change any default behavior?

Not from an end-user perspective. 

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

Feature is not intended to be a toggleable. However, each distribution can decide for themselves.

Note: While enabling the feature on a live cluster is easier (owing to custom logic in ExternalJWTSigner's implementation), disabling might not make for a good experience because there will be JWTs signed by external keys that are still hanging around. Disabling will either require the supported external keys to be moved onto the disk and be configured via `--service-account-key-file` or the respective distribution will need to come with a fair warning that all active service account tokens will get invalidated.

###### What happens if we re-enable the feature if it was previously rolled back?

experience would be the same as when enabling.

###### Are there any tests for feature enablement/disablement?

<<[UNRESOLVED]>>

> Feature is not intended to be a toggleable

Reviewer 4: This is a non-starter.  You always have to be able to go back, and in this case I don't even this it is that hard to go back since you just need the public key for the externally signed tokens to keep verifying them.

Author: It is possible to achieve; Reworded with hints on how to...

<<[/UNRESOLVED]>>

### Rollout, Upgrade and Rollback Planning

Since it is an opt-in feature; upgrading existing clusters to versions with this feature shall be a no-op in context of the feature.

###### How can a rollout or rollback fail? Can it impact already running workloads?

Since it is an opt-in feature, existing clusters shall remain unaffected. However, since we are touching paths for JWT signing and authentication, if a bug slips in, it can disrupt the respective operation.

###### What specific metrics should inform a rollback?

for a rolled-up (upgraded) cluster:
- Increase in error rates on service account token creation operation.

> - sudden increase in PERMISSION_DENIED errors.

Reviewer 4: This isn't specific enough.

Author: named specific metrics to look out for.

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

<!-- TODO
Describe manual testing that was done and the outcomes.
Longer term, we may want to require automated upgrade/rollback tests, but we
are missing a bunch of machinery and tooling and can't do that now.
-->

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

No.

> ### Monitoring Requirements

Author: Waiting until beta to release to fill it out. Let me know if you think we need it sooner?

> ### Monitoring Requirements

Author: have filled the parts that I could. Would add the remaining after starting implementation.

- ExternalJWTSigner
    - Usage description:
      This service will act as a client to the external key management solution that a distribution will integrate with.

      NOTE: this being a configurable/opt-in feature, the impact will only be seen on clusters that are using the feature.

      - Impact of its outage on the feature:
        - kube-apiserver will not be able to sign new service account JWTs.
        - old JWTs will continue to be supported since we have an in cluster cache.
        - refreshing supported key sets will not be possible.
      - Impact of its degraded performance or high-error rates on the feature:
        - service account token generation requests might require retries.
        - issues when syncing support key sets might cause intermediate auth failures only if there are changes in supported key sets.

### Scalability

###### Will enabling / using this feature result in any new API calls?

- Sign:
  - This will be a call from kube-apiserver to an external service account JWT signer.

>   - The call will be made every time kube-apiserver receives a service account token request.

Author: Done

  - QPS will vary according to the cluster usage.
  - ExternalJWTSigner's throughput will vary from distribution to distribution.

- FetchKeys:
  - This is again a call from kube-apiserver to external service account JWT signer.
  - It will be a periodic call to keep the OIDC provider in sync with the supported keys; Frequency will be decide by `suggested_next_fetch_at` returned by external signer. 
  - It will also be a dynamic call whenever kube-apiserver comes across a token signed by a key_id that it does not recognize.
  - QPS will be rather low since:
    - the signing keys shall not change very frequently.
    - even a small value of `suggested_next_fetch_at` will result in less than 1 qps.
  - ExternalJWTSigner's throughput will vary from distribution to distribution.

###### Will enabling / using this feature result in introducing new API types?

No new K8s api.

###### Will enabling / using this feature result in any new calls to the cloud provider?

> - Create toke

Author: Done

###### What are other known failure modes?

<!-- TODO
For each of them, fill in the following information by copying the below template:
  - [Failure mode brief description]
    - Detection: How can it be detected via metrics? Stated another way:
      How can an operator troubleshoot without logging into a master or worker node?
    - Mitigations: What can be done to stop the bleeding, especially for already
      running user workloads?
    - Diagnostics: What are the useful log messages and their required logging
      levels that could help debug the issue?
      Not required until the feature graduated to beta.
    - Testing: Are there any tests for failure mode? If not, describe why.
-->

###### What steps should be taken if SLOs are not being met to determine the problem?

The improvement will likely need to be made on respective cloud-providers API. Confirming if it's the source of the problem can be done via. respective external metrics.

## Implementation History

Initial PRs: 
- kubernetes/kubernetes#73110
- kubernetes/kubernetes#125177

## Drawbacks

> There seem to be no drawbacks per say. It's an opt-in feature and does not interfere with existing functionality. If anything, it should be implemented and improved upon if need be.

Author: reworded and called it out.

  - [Non-Goals](#non-goals)
- [Proposal](#proposal)
  - [Preserve existing behavior](#preserve-existing-behavior)
  - [New API](#new-api)
  - [Plugins] (#plugins)

>   - [Updates for token generation](#updates-for-token-generation)

Author: done

  - [Updates for supported public keys](#updates-for-supported-public-keys)
  - [Notes/Constraints/Caveats](#notesconstraintscaveats)
  - [Risks and Mitigations](#risks-and-mitigations)
- [Design details](design-details)
  - [ExternalJWTSigner RPC](#externaljwtsigner-rpc)
  - [Test Plan](#test-plan)
	  - [Prerequisite testing updates](#prerequisite-testing-updates)
	  - [Unit tests](#unit-tests)
  - [Graduation Criteria](#graduation-criteria)
  - [Upgrade/Downgrade Strategy](#upgradedowngrade-strategy)
  - [Scalability](#scalability)
- [Implementation History](#implementation-history)
<!-- /toc -->

## Summary

> Service account keys are used by kube-apiserver for JWT signing and authentication. Keys are stored on the disk and are loaded by kube-apiserver during process start time. Key management for service account keys is entirely within the cluster and is less flexible. This KEP proposes updates that will allow Kubernetes distributions to integrate with key management solutions of their choice (eg: HSMs, cloud KMSes).  

Reviewer 3: nit:

> "... and is less flexible." 

Less flexible than what? It might be better to either compare it to something, or just state what it can't/doesn't do (ex: keys are exportable)

Author: reworded to `...is less flexible in that the signing keys remain same for the lifetime of the process`

## Motivation

Enable integration with alternate key management solutions for service account keys. While at it, make the association between kube-apiserver and service account keys more agile.

> Benefits are 2 fold:

Author: done

1. Ease of rotation.
  At present, kube-apiserver loads service account keys at process start time and the key remains the same for the life-time of the process. Thus, rotating keys require kube-apiserver restart. External key management eliminates the need to restart.
2. Enhanced security.

>   kube-apiserver loads service account keys from a file on the disk. Which means that the service account keys are always present on the disk and can be easily accessed by the cluster administrators. With an external signer, we do not need to store the service account keys on the disk. 

Reviewer: Maybe rephrase as

"This means anyone with privileged access to the kube-apiserver configuration / filesystem could exfiltrate signing material. An external signer that never returns signing material would mitigate this risk."

Author: done

### Goals

- Enable integration with HSMs and cloud KMSes
- Support out-of-process JWT signing
- Support listing public verifying keys

> - Prevent external signers from controlling the claims embedded in JWT

Reviewer: ```suggestion
- Ensure tokens signed by external signers remain consistent with tokens issued by kube-apiserver (consistent headers, algorithms, claims, etc) 
```

- Preserve existing behavior and performance for keys not read over a socket
<<[UNRESOLVED]>>
- Move legacy token signing from kube-controller-manager to kube-apiserver and thus external signing.
<<[/UNRESOLVED]>>

### Non-Goals

- Reading TLS serving certificates and key from a socket or reloading of kube-apiserver with new cert and key
- Reading any other certificates from a file

## Proposal

### Preserve existing behavior

The kube-apiserver flags `--service-account-key-file` and `--service-account-signing-key-file` will continue to be used for reading from files.

### New API

> I'm proposing creating a new versioned grpc API (ExternalJWTSigner) under `k8s.io/kubernetes/pkg/serviceaccount`. This will be similar to how the KMS envelope encryption has an API at `k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/v1beta1/service.proto`. It will have methods to sign JWTs and fetch supported keys (proto in [Design Details](#design-details) section). 

Reviewer: ```suggestion
A new versioned grpc API (ExternalJWTSigner) will be created under `k8s.io/kubernetes/pkg/serviceaccount`. This will be similar to how the KMS envelope encryption has an API at `k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/v1beta1/service.proto`. It will have methods to sign JWTs and fetch supported keys (proto in [Design Details](#design-details) section). 
```

### Plugins

- A signer plugin will be a client to the new grpc API thus providing support for external signing.
- A public key cache will integrate with the same grpc to fetch supported public keys.

### Updates for token generation

> As of Kubernetes `v1.13.2`, The Kubernetes Control Plane config accepts `ServiceAccountIssuer` which is an interface of type `serviceaccount.TokenGenerator` as defined in `pkg/serviceaccount/jwt.go`. This enables defining alternate implementations of token generators. Thus, the default TokenGenerator can be replaced by one that integrates with ExternalJWTSigner.

Reviewer: can simplify to just link to the interface, describe the current static file implementation in a sentence, and indicate we'd have an alternate implementation backed by the grpc client

Author: Done

### Updates for supported public keys

Starting ` v1.31` (via. the [PR](https://github.com/kubernetes/kubernetes/pull/125177) by liggit@), `ServiceAccountPublicKeys` in kube-apiserver config was replaced by `ServiceAccountPublicKeysGetter` interface. The static keys held by `ServiceAccountPublicKeys` were hence moved to be managed by the implementation called `StaticPublicKeysGetter`. This presents an opportunity to add a new implementation, one that will get public keys from ExternalJWTSigner depending on how the cluster is configured.

> In the same PR, `OpenIDMetadataServer` was updated to accept implementation of `OpenIDMetadataProvider` as defined in  `pkg/serviceaccount/openidmetadata.go`. `OpenIDMetadataProvider` interacts with `ServiceAccountPublicKeysGetter` for serving supported keys which enables flexible replacing `StaticPublicKeysGetter` with a dynamic key getter (will integrate with ExternalJWTSigner) depending on how cluster is configured.

Reviewer: can simplify to drop history and PR links, just link to the current interface(s), describe the current static file implementations in a sentence, and indicate we'd have an alternate implementation backed by the grpc client and key cache

Author: done

### Notes/Constraints/Caveats

> - A new flag `--key-service-url` will be added to kube-apiserver specifying a unix socket where the key service will be accessible.

Reviewer: consider naming this `...-endpoint` to be consistent with how we refer to the KMS GRPC endpoint

probably want to prefix with `--service-account-...` to make it clear this is service-account related and groups with the other service account flags

Suggest `--service-account-signing-endpoint`

Author: Done

- The flag `--key-service-url` can either be set to the location of a UDS on a filesystem, or be prefixed with an @ symbol and name a UDS in the abstract socket namespace.

> - combining local and external keys will not be supported.

Reviewer: ```suggestion
- combining local and external keys will not be supported inside kube-apiserver. This *could* be achieved inside an external signer implementation by combining local public keys with remote public keys and remote signing requests to accommodate a migration.
```

- Specifying both, `--service-account-key-file` and `--key-service-url` shall result in an error.

### Risks and Mitigations

> - New token generation and validation could suffer a performance difference when reading over a socket, as an external process will be signing data.

Reviewer: We need to benchmark / document the overhead inherent in the grpc call out (can benchmark a test implementation that just moves the current signing into a grpc test service). Then it is up to implementers to benchmark their own signers and decide if the resulting latency is acceptable.

Author: have added a Risk/Mitigation pair.

> - New token generation and validation could suffer a performance difference when reading over a socket, as an external process will be signing data.

Author: Reworded

> - Signing and verifying tokens over a grpc API carries the risk of a server side request forgery, where a malicious client could generate tokens. To mitigate this risk, the API will only be accessible over a unix socket.

Reviewer: ```suggestion
- Signing and verifying tokens over a grpc API carries the risk of a server side request forgery, where a malicious client could generate tokens. To mitigate this risk, the API will only be accessible over a unix socket protected by standard file permissions. Document that the socket should be limited to privileged or specific uids.
```

> - External signers will only return signature and header while the token will be assembled by TokenGenerator implementation for external signing. This will help kube-apiserver control the payload and thus the claims.

Reviewer: reword this as a risk that this sentence is the mitigation to (risk is token headers and claims will diverge, mitigation is that kube-apiserver verifies the header algorithm is acceptable, and controls the claims)

Author: Done

## Design Details

### ExternalJWTSigner RPC

Will be served on a socket as configured via the `--key-service-url` flag.

```proto
syntax = "proto3";

// This service is served by a process on a local Unix Domain Socket.
service ExternalJWTSigner {
  // Sign takes a serialized JWT payload, and returns the serialized header and
  // signature.  The caller can then assemble the JWT from the header, payload,
  // and signature.
  //
  // The plugin MUST set a key id in the returned JWT header.
  rpc Sign(SignJWTRequest) returns (SignJWTResponse) {}

  // FetchKeys returns the set of public keys that are trusted to sign
  // Kubernetes service account tokens. Kube-apiserver will call this RPC:
  //
  // * Every time it tries to validate a JWT carrying an unknown key ID, and
  //
  // * Periodically, so it can serve reasonably-up-to-date keys from the OIDC
  //   JWKs endpoint.
  rpc FetchKeys(FetchKeysRequest) returns (FetchKeysResponse) {}

> }

Reviewer: add `<<UNRESOLVED>>` section for the meta / version / health bit here

Author: Done

message SignJWTRequest {

>   // The raw payload to be signed.  A serialized JSON object, no base64 wrapping.

Reviewer: let's make this already be base64 encoded to remove a possible mistake in the external signer encoding (and symmetric with the "Already wrapped in URL-safe base64" fields in the response)

Author: Done

>   bytes payload = 1;

Reviewer: let's call this `claims` to make it clear which bit of the jwt this is

```suggestion
  bytes claims = 1;
```

Author: Done

}

message SignJWTResponse {
  // The header for the JWT.  Already wrapped in URL-safe base64, valid for

>   // directly inserting into the final JWT.

Reviewer: hoist in the docs from https://docs.google.com/document/d/1QVtBX8J2Tk70toefDegXMXigMFzrdQfezjlhBrKR29Y/edit?tab=t.0#bookmark=id.xf2j5nbnuczz with the restrictions kube-apiserver will put on the returned header

```
  // header must contain alg, kid, typ claims.
  // type must be “JWT”.
  // alg must be one of the algorithms supported by kube-apiserver (currently RS256, ES256, ES384, ES512).
```

Author: Done

  bytes header = 1;

  // The signature for the JWT.  Already wrapped in URL-safe base64, valid for
  // directly inserting into the final JWT.
  bytes signature = 2;
}

message FetchKeysRequest {}

message FetchKeysResponse {
  repeated Key keys = 1;

  // The timestamp when this data was pulled from the authoritative source of
  // truth for verification keys.
  google.protobuf.Timestamp data_timestamp = 2;

  // The timestamp for when it's appropriate to refetch verification keys to 
  // pick changes if any.

>   google.protobuf.Timestamp suggested_next_fetch_at = 3;

Reviewer: is there a minimum we'd accept here? (what would kube-apiserver do if this is before time.Now()? make sure we don't hot-loop or do something crazy if skew is encountered like that)

is this what informs the GetCacheAgeMaxSeconds() implementation for PublicKeysGetter? something like `suggested_next_fetch_at.Sub(time.Now())/time.Second`? that would mean the OIDC docs would return shrinking cache maxAge headers as we got closer and closer to the internal refresh time

Need to document how to use this and/or the metrics to safely add new keys and promote them to signing keys in an external signer

Author: the maxAge is eventually being used in cache control: https://github.com/kubernetes/kubernetes/blob/a15dca3765d1646d6c29832fdfe1a942e310f7a4/pkg/routes/openidmetadata.go#L95

if it slips to a negative value, it can cause a hot-loop. Makes me wonder if OIDCMetadata server shall at all depend on this field? If feels like a better option would be to have a fixed refresh interval for OIDC Metadata. Having said that, maybe we can omit out `suggested_next_fetch_at` (atleast to begin with)?

Reviewer: what is the surface area of the JWK schema? I don't know offhand how sprawling / flexible the JWK schema is, and it seems like whatever could be returned from the external signer would would have to be plumbed all the way through kube-apiserver and published in oidc discovery... my goals for the initial iteration of this are to make the surface area of the issued tokens and published oidc discovery *identical* whether you're using an internal or external signer

Author: Done

}

message Key {
  // A unique identifier for this key.
  string key_id = 1;

  // The public key, PKIX-serialized.
  bytes key = 2;
}
```

### Test Plan

[x] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

<!-- TODO -->

##### Unit tests

Will need coverage for:

- pkg/kubeapiserver/options/authentication.go
- the signer plugin 
- the key cache plugin

<!-- TODO
Add info on concerned test in below form as read from https://testgrid.k8s.io/sig-testing-canaries#ci-kubernetes-coverage-unit

- `<package>`: `<date>` - `<test coverage>`
 -->

> 

Reviewer: add an integration test section describing wiring up a real working kube-apiserver configured to run with this signer (KMS KEP might have a good testing section you can mimic)

Author: Added some initial bits. Can you point me to the KEP that you are talking about?

Reviewer: https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/3299-kms-v2-improvements#integration-tests

### Graduation Criteria

<!-- TODO -->

### Upgrade/Downgrade Strategy

Existing clusters preserving existing functionality: 
- Need no changes.

Existing clusters using enhanced functionality: 
- Need a MasterVM workload that implements ExternalJWTSigner
- Need to configure the cluster with the `--key-service-url` flag. 

### Scalability

<!-- TODO -->

## Implementation History

Initial PRs: 
- kubernetes/kubernetes#73110

> - kubernetes/kubernetes#125177

Author: Done

  - [Non-Goals](#non-goals)
- [Proposal](#proposal)
  - [Preserve existing behavior](#preserve-existing-behavior)
  - [New API](#new-api)
  - [Plugins] (#plugins)
  - [Updates to API server token generation](#updates-to-api-server-token-generation)
  - [Updates for supported public keys](#updates-for-supported-public-keys)
  - [Notes/Constraints/Caveats](#notesconstraintscaveats)
  - [Risks and Mitigations](#risks-and-mitigations)
- [Design details](design-details)
  - [KeyService RPC](#keyservice-rpc)
  - [Test Plan](#test-plan)
	  - [Prerequisite testing updates](#prerequisite-testing-updates)
	  - [Unit tests](#unit-tests)
  - [Graduation Criteria](#graduation-criteria)
  - [Upgrade/Downgrade Strategy](#upgradedowngrade-strategy)
  - [Scalability](#scalability)
- [Implementation History](#implementation-history)
<!-- /toc -->

## Summary

>  Kubernetes API server uses service account keys for signing JWT. Signing keys are stored on the disk and are loaded by Api server as the process starts. Key management is entirely within the cluster and is less flexible. It will be of advantage to provide a configurable capability that allows external JWT signing and key management.

Author: Done and Done.

## Motivation

KEP proposes external JWT signing and listing of supported public keys.

Benefits are 2 fold:
1. Ease of rotation.
  At present, ApiServer loads service account keys at process start time and the key remains the same for the life-time of the process. Thus, rotating keys require ApiServer restart. External key management eliminates the need to restart.
2. Enhanced security.
  ApiServer loads service account keys from a path on the disk. Which means that the service account keys are always present on the disk and can be easily accessed by the cluster administrators. With an external signer, we need not store the service account key on disk. 

### Goals

> - Support for out-of-process JWT signing

Reviewer 6: We should explicitly call out that we want to make it possible to support HSMs and cloud KMSes.

Additionally, we should have as a goal that external signer implementations should not be able to introduce custom claims into the JWTs

Author: Done and Done

- Support for listing public verifying keys
- Preserve existing behavior and performance for keys not read over a socket

> - Move legacy token signing from kube-controller-manager to Kubernetes ApiServer and thus external signing.

Reviewer 6: I think we're undecided on this point.  Maybe put it in an <<[UNRESOLVED]>> block.

Author: Wrapped under UNRESOLVED block

### Preserve existing behavior

The API server flags `--service-account-key-file` and `--service-account-signing-key-file` will continue to be used for reading from files.

### New API

> I'm proposing creating a new versioned grpc API (KeyService) under `k8s.io/kubernetes/pkg/serviceaccount`. This will be similar to how the KMS envelope encryption has an API at `k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/v1beta1/service.proto`. It will have methods to Sign and JWT and Fetch supported keys (proto in Details section). 

Reviewer 6: "Sign and JWT" --- "Sign JWTs"?

Author: Oops! corrected.

### Plugins

> - A signer plugin will be a client to the new grpc API thus providing support for external Signing.

Reviewer 6: Nit: signing should not be capitalized.

Author: Done

> - A public key Cache will integrate with the same grpc to fetch supported public keys.

Reviewer 6: Nit: Cache should not be capitalized.

### Updates for supported public keys

Starting ` v1.31` (via. the [PR](https://github.com/kubernetes/kubernetes/pull/125177) by liggit@), `ServiceAccountPublicKeys` in ApiServer config was replaced by `ServiceAccountPublicKeysGetter` interface. The static keys held by `ServiceAccountPublicKeys` were hence moved to be managed by the implementation called `StaticPublicKeysGetter`. This presents an opportunity to add a new implementation, one that will get public keys from KeyService depending on how the cluster is configured.

In the same PR, `OpenIDMetadataServer` was updated to accept implementation of `OpenIDMetadataProvider` as defined in  `pkg/serviceaccount/openidmetadata.go`. `OpenIDMetadataProvider` interacts with `ServiceAccountPublicKeysGetter` for serving supported keys which enables flexible replacing `StaticPublicKeysGetter` with a dynamic key getter (will integrate with KeyService) depending on how cluster is configured.

### Notes/Constraints/Caveats

> - A new flag `--key-service-url` will be added to the API server specifying a unix socket where the key service will be accessible.

Reviewer 6: Specify that the value can either name the location of a UDS on a filesystem, or be prefixed with an `@` symbol and name a UDS in the abstract socket namespace.

Author: done

> - cumulation on local and external keys will not be supported.

Reviewer 6: "cumulation" --> "combining"?

Author: done

> - If Both, `--service-account-key-file` and `--key-service-url` are specified, keys from external signers will override the keys at `service-account-key-file` as supported public keys.

Reviewer 6: We should probably just throw an error if this happens?

Also, "Both" should not be capitalized.

Author: done and done

### Risks and Mitigations

- New token generation and validation could suffer a performance difference when reading over a socket, as an external process will be signing data.
- Signing and verifying tokens over a grpc API carries the risk of a server side request forgery, where a malicious client could generate tokens. To mitigate this risk, the API will only be accessible over a unix socket.
- External signers will only return signature and header while the token will be assembled by TokenGenerator implementation for external signing. This will help ApiServer control the payload and thus the claims.

## Design Details

### KeyService RPC

> will be served on a socket as configured via the `--key-service-url` flag.

Reviewer 6: "Will"

Author: Done

```proto
// This service is served by a process on a local Unix Domain Socket.

> service KeyService {

Reviewer 6: KeyService is very generic, and could also describe the KMS plugin for etcd storage.  We should use a more specific name, like "KubernetesExternalJWTSigner", or something like that.

Author: calling it `ExternalJWTSigner`

  // Sign takes a serialized JWT payload, and returns the serialized header and
  // signature.  The caller can then assemble the JWT from the header, payload,
  // and signature.
  //
  // The plugin MUST set a key id in the returned JWT header.
  rpc Sign(SignJWTRequest) returns (SignJWTResponse) {}

  // FetchKeys returns the set of public keys that are trusted to sign
  // Kubernetes service account tokens. Kube-apiserver will call this RPC:
  //
  // * Every time it tries to validate a JWT carrying an unknown key ID, and
  //
  // * Periodically, so it can serve reasonably-up-to-date keys from the OIDC
  //   JWKs endpoint.
  rpc FetchKeys(FetchKeysRequest) returns (FetchKeysResponse) {}
}

message SignJWTRequest {
  // The raw payload to be signed.  A serialized JSON object, no base64 wrapping.
  optional bytes payload = 1;
}

message SignJWTResponse {
  // The header for the JWT.  Already wrapped in URL-safe base64, valid for
  // directly inserting into the final JWT.

>   optional bytes header = 1;

Reviewer 6: Is this using proto2 or proto3?  If proto3, we should not specify optional unless we specifically need presence information (which I don't believe we do).

Author: made it explicit that it's proto3. removed optional.

  // The signature for the JWT.  Already wrapped in URL-safe base64, valid for
  // directly inserting into the final JWT.
  optional bytes signature = 2;
}

message FetchKeysRequest {

>     optional string triggering_key_id = 1;

Reviewer 6: Consensus in the discussion was to drop this field until there was a concrete use case.

Author: done

}

message FetchKeysResponse {
  repeated Key keys = 2;

  // The timestamp when this data was pulled from the authoritative source of
  // truth for verification keys.
  optional google.protobuf.Timestamp data_timestamp = 3;

> }

Reviewer 6: We also wanted to add a `suggest_next_fetch_at` timestamp field, to allow the plugin to control the polling frequency.

Author: Added
